x/delegation/keeper/share.go

```diff
@@ -0,0 +1,210 @@
+// deviation caused by precision, especially when a large number of restakers undertake
+// undelegation. Additionally, the last undelegation from an operator will undelegate all
+// remaining token to avoid the calculated token amount is bigger than the remaining token
+// caused by the bankers rounding.
+func TokensFromShares(stakerShare, totalShare sdkmath.LegacyDec, operatorAmount sdkmath.Int) (sdkmath.Int, error) {
+	if totalShare.IsZero() {
+		return sdkmath.NewInt(0), delegationtypes.ErrDivisorIsZero
+	}
+	return (stakerShare.MulInt(operatorAmount)).Quo(totalShare).TruncateInt(), nil
+}
+
+// SharesFromTokens returns the shares of a delegation given a delegated amount. It
+// returns an error if the validator has no tokens.
+// It uses `LegacyDec.QuoInt` to calculate the quotient, the calculated result will
+// be truncated through the truncation implemented by golang's standard big.Int.
+// So the calculated share might tend to be smaller, but it seems acceptable, because
+// we need to make sure the staker can't get a bigger share than they should get.
+func SharesFromTokens(totalShare sdkmath.LegacyDec, stakerAmount, operatorAmount sdkmath.Int) (sdkmath.LegacyDec, error) {
+	if operatorAmount.IsZero() {
+		return sdkmath.LegacyZeroDec(), delegationtypes.ErrDivisorIsZero
+	}
+	return totalShare.MulInt(stakerAmount).QuoInt(operatorAmount), nil
+}
+
+// CalculateShare calculates the S_j
+// S_j = S * T_j / T, `S` and `T` is the current asset share and amount of operator,
+// and the T_j represents the change in staker's asset amount when some external
+// operations occur, such as: delegation, undelegation and non-instantaneous slashing.
+// A special case is the initial delegation, when T = 0 and S = 0, so T_j / T is undefined.
+// For the initial delegation, delegator j who delegates T_j tokens receive S_j = T_j shares.
+func (k Keeper) CalculateShare(ctx sdk.Context, operator sdk.AccAddress, assetID string, amount sdkmath.Int) (share sdkmath.LegacyDec, err error) {
+	// get the total share of operator
+	isExist := k.assetsKeeper.IsOperatorAssetExist(ctx, operator, assetID)
+	var info *assetstype.OperatorAssetInfo
+	if isExist {
+		info, err = k.assetsKeeper.GetOperatorSpecifiedAssetInfo(ctx, operator, assetID)
+		if err != nil {
+			return share, err
+		}
+	}
+	if !isExist || info.TotalShare.IsZero() {
+		// the first delegation to a validator sets the exchange rate to one
+		share = sdk.NewDecFromInt(amount)
+	} else {
+		share, err = SharesFromTokens(info.TotalShare, amount, info.TotalAmount)
+		if err != nil {
+			return share, err
+		}
+	}
+	return share, nil
+}
+
+// ValidateUndeleagtionAmount validates that a given undelegation amount is
+// valid based on upon the converted shares. If the amount is valid, the total
+// amount of respective shares is returned, otherwise an error is returned.
+func (k Keeper) ValidateUndeleagtionAmount(
```

Comment:
Typo: `ValidateUndelegationAmount`

Reply:
Thanks, it will be fixed in the next PR.